### PR TITLE
Prevent multiple restarts when changing multiple attributes on a

### PR DESCRIPTION
**Pull request by josejulio**
composite operation.

I like this approach because I don't need to copy code from parent class.

@jmazzitelli please take a look at let me know if I should use this PR or the other as [base](https://github.com/hawkular/hawkular-agent/pull/291).

I'll continue on Monday, will add itests to check if the agent is being restarted more than once.

### CODE REVIEW

File: hawkular-wildfly-agent/src/main/java/org/hawkular/agent/monitor/extension/MonitorServiceRestartParentAttributeHandler.java [hunk 1]
```diff
@@ -58,15 +65,39 @@ protected boolean requiresRuntime(OperationContext context) {
     @Override
     protected void removeServices(OperationContext context, ServiceName parentService, ModelNode parentModel)
```
Comment (josejulio) [inline hawkular-wildfly-agent/src/main/java/org/hawkular/agent/monitor/extension/MonitorServiceRestartParentAttributeHandler.java:67]: Let me know if you would like to have this on Util (or somewhere else) to execute blocks of code only once in a `OperationContext`.

Comment (jmazzitelli) [reply hawkular-wildfly-agent/src/main/java/org/hawkular/agent/monitor/extension/MonitorServiceRestartParentAttributeHandler.java:67]: I think this is good as is (assuming it all works :)) No need to put this in Util yet. Don't think we'll have a need for it - but if we do, we can refactor then.

I assume this works on EAP 6.4 as well as 7.0 and 7.1?

Let's see if we can get an itest in for this before merging. 

@pilhuhn  asked for an agent release today so we'll need to get something merged today to fix this problem before we release.

**Comment by jshaughn** [comment]
Testing this locally the changes seems to work, so good job so far.  But the bad news is that because the restart is automatic the agent Command (in my test, UpdateCollectionIntervalsCommand) can not get its response sent back, because the agent is shutting down.

We need to have a way to get a Response to the client or the client won't know the command succeeded.

File: hawkular-wildfly-agent/src/main/java/org/hawkular/agent/monitor/extension/MonitorServiceRestartParentAttributeHandler.java (hunk 1 above)
Comment (josejulio) [reply hawkular-wildfly-agent/src/main/java/org/hawkular/agent/monitor/extension/MonitorServiceRestartParentAttributeHandler.java:67]: EAP6.4 seems to be upset with something at this point. I'm trying to reason with it.

**Comment by josejulio** [comment]
Umh, maybe the shutdown should wait for all the current jobs to finish or something like that?

**Comment by jmazzitelli** [comment]
The problem is the restart is going to be completely separate from the command gateway command processing. We are getting commands in from the WebSocket, performing these DMR operations (which trigger this restart), but before the WebSocket processing receives the DMR operation response so it can send back the WebSocket response, the agent is restarted (which kills the WebSocket!).

Essentially, its a suicide mission - and the WebSocket processing kills itself before it knows it can send a response.

We need somehow to relay information from the DMR operation to the WebSocket command processor.

**Comment by jmazzitelli** [comment]
@josejulio this is not to say you shouldn't get this PR working (with EAP 6 and 7) and eventually get it merged. The problem will need to be for us to workaround the WebSocket issue. It may mean enhancing the code in this PR, but this code in some form is still going to be needed.

**Comment by pilhuhn** [comment]
On 13 Feb 2017, at 16:53, John Mazzitelli wrote:

> We need somehow to relay information from the DMR operation to the 
> WebSocket command processor.

Is there a way to dry-run the (composite) operation and
then send a response back upon the result?

That may be either failure = would fail for sure
or Looks good = it will most likely succeed.

And then we cache the information that we are running
the operation, so that after the restart we can send
a success message when we see the cached info

**Comment by josejulio** [comment]
@jmazzitelli I got it to run on EAP6.4. 
Before (on EAP6.4) we needed to mark an step as Completed, thus. Will test on eap7.1 and push the code.

**Comment by jshaughn** [comment]
@josejulio @jmazzitelli , I think I have a solution, it requires one change to this PR.  Currently We have this in `MonitorServiceRestartParentAttributeHandler:`

```
@Override
protected boolean isResourceServiceRestartAllowed(OperationContext context, ServiceController<?> service) {
    return true;
}
```

I think we want to remove this and go back to the default impl in the superclass.  What that means is that we don't actually perform a restart by default even if the operations perform changes that require a restart to be picked up.  It puts the responsibility on the caller to supply a header saying it's allowed.  For me this is a good thing, I don't really think we should always immediately restart, it could be the case that  the restart should be deferred.

To have you operation perform the automated restart you supply the operation header `ModelDescriptionConstants.ALLOW_RESOURCE_SERVICE_RESTART=true`.  This can be applied manually or with `OperationBuilder.allowResourceServiceRestart()`.

So that lets you perform an operation and control whether the restart kicks in.   This still leaves one problem, like we have in the new UpdateCollectionIntervals command.  Here we want a restart but because we are restarting the agent itself, the restart has to wait for the command response to be returned before getting invoked.  An immediate restart will prevent the response from being sent.  So, in this case we don't supply the `ALLOW_RESOURCE_SERVICE_RESTART` and instead invoke the restart manually from the command.  But we need it to be deferred.  This requires a small change in the agent's start command impl, which I have in a separate PR, but basically it puts the restart logic into a thread and offers a new "delay" attribute for the operation.

If you guys are OK with this I think we can move forward.  Either merge this as-is and I'll remove the  `isResourceServiceRestartAllowed()` override in my command PR, or update this first.  Otherwise, let me know what I missed...

**Comment by jmazzitelli** [comment]
I'm going to merge this as-is. So we will have a baseline for this working "restart service" code.

We'll then work on the second PR that is needed for the update interval stuff.

**Pull request merged**